### PR TITLE
Adds types to Lifecycle Objects

### PR DESCRIPTION
**Pull request by InvincibleRMC**
Builds off of #1235.

### CODE REVIEW

File: rclpy/rclpy/lifecycle/node.py [hunk 1]
```diff
@@ -273,30 +317,52 @@ def on_error(self, state: LifecycleState) -> TransitionCallbackReturn:
...
+        if publisher_class:
             raise TypeError(
                 "create_publisher() got an unexpected keyword argument 'publisher_class'")
-        pub = Node.create_publisher(self, *args, **kwargs, publisher_class=LifecyclePublisher)
+        pub = Node.create_publisher(self, msg_type, topic, qos_profile,
+                                    publisher_class=LifecyclePublisher,
+                                    **kwargs)
+
+        if not isinstance(pub, LifecyclePublisher):
+            raise Exception('Node failed to create LifecyclePublisher.')
```
Comment (sloretz) [inline rclpy/rclpy/lifecycle/node.py:347]: nit, please use RuntimeError

**Review by sloretz — CHANGES_REQUESTED** [review]
LGTM with one nitpick!

**Review by sloretz — APPROVED** [review]

**Comment by sloretz** [comment]
@mergifyio update

**Comment by sloretz** [comment]
Pulls: ros2/rclpy#1338
Gist: https://gist.githubusercontent.com/sloretz/5ae8c9493cbb2f6cd304e97a2d4ef14c/raw/0cc08b8cc6fe2a9191ebc2969ec1ba7b33589bc5/ros2.repos
BUILD args:  --packages-above-and-dependencies rclpy
TEST args:  --packages-above rclpy
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/14473
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/21743/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/16017/)
* Linux-rhel [Build Status](http://ci.ros2.org/job/ci_linux-rhel/1248/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/22446/)

**Review by sloretz — APPROVED** [review]

**Comment by sloretz** [comment]
Pulls: ros2/rclpy#1338
Gist: https://gist.githubusercontent.com/sloretz/26afab0411e269cd97f9c479b0507b67/raw/0cc08b8cc6fe2a9191ebc2969ec1ba7b33589bc5/ros2.repos
BUILD args:  --packages-above-and-dependencies rclpy
TEST args:  --packages-above rclpy
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/14474
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/21744/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/16018/)
* Linux-rhel [Build Status](http://ci.ros2.org/job/ci_linux-rhel/1249/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/22447/)

**Comment by sloretz** [comment]
@InvincibleRMC Mind fixing the conflict in this one?

**Comment by InvincibleRMC** [comment]
@sloretz The conflicts have been fixed.

**Comment by fujitatomoya** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/21945/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/16212/)
* Linux-rhel [Build Status](http://ci.ros2.org/job/ci_linux-rhel/1462/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/22672/)

**Review by ahcorde — APPROVED** [review]

**Pull request merged**